File: 2023/Python/Day9.py
# -*- coding: utf-8 -*-
"""
Created on Sat Dec  9 06:44:22 2023

@author: jkv
"""

class HistoryExtrapolator:
    def __init__(self, history):
        self.history = history
        self.sequences = self._generate_difference_sequences()

    def _generate_difference_sequences(self):
        sequences = [self.history]

        while sequences[-1].count(0) != len(sequences[-1]):
            new_sequence = [sequences[-1][i+1] - sequences[-1][i] for i in range(len(sequences[-1])-1)]
            sequences.append(new_sequence)

        return sequences

    def extrapolate(self, direction='next'):
        for i in range(len(self.sequences) - 2, -1, -1):
            if direction == 'next':
                self.sequences[i].append(self.sequences[i][-1] + self.sequences[i+1][-1])
            elif direction == 'previous':
                self.sequences[i].insert(0, self.sequences[i][0] - self.sequences[i+1][0])

        if direction == 'next':
            return self.sequences[0][-1]
        elif direction == 'previous':
            return self.sequences[0][0]
        
    def load_data(file_path):
        with open(file_path, 'r') as file:
            data = file.readlines()
        return [list(map(int, line.split())) for line in data]

# Load the data
file_path = 'day9.txt'
histories = load_data(file_path)


part1 = 0
part2 = 0
for history in histories:
    extrapolator = HistoryExtrapolator(history)
    next_value = extrapolator.extrapolate('next')
    prev_value = extrapolator.extrapolate('previous')
    part1 += next_value
    part2 += prev_value

part1, part2

File: 2023/R/Day9.R
library(tidyverse)

# Func to generate diffs in sequences
diff_sequences <- function(history) {
  sequences <- list(history)
  while (length(unique(sequences[[length(sequences)]])) > 1) {
    new_sequence <- diff(sequences[[length(sequences)]])
    sequences <- append(sequences, list(new_sequence))
  }
  return(sequences)
}

# Function to extrapolate the next or previous value
extrapolate_value <- function(history, direction = "next") {
  sequences <- diff_sequences(history)
  if (direction == "previous") {
    sequences <- lapply(sequences, rev)
  }
  
  for (i in length(sequences):2) {
    sequences[[i - 1]] <- c(sequences[[i - 1]], tail(sequences[[i - 1]], n=1) + tail(sequences[[i]], n=1))
  }
  
  if (direction == "next") {
    return(tail(sequences[[1]], n=1))
  } else {
    return(head(sequences[[1]], n=1))
  }
}

# Function to load data and calculate the sum of extrapolated values
calculate_sum <- function(file_path, direction = "next") {
  histories <- read_lines(file_path) %>% 
    map(~str_split(.x, " ") %>% unlist() %>% as.numeric())
  
  sum_values <- map_dbl(histories, ~extrapolate_value(.x, direction))
  return(sum(sum_values))
}

# Load the data and calculate the sums for both parts
file_path <- "day9.txt"
sum_next_values <- calculate_sum(file_path, "next")
sum_previous_values <- calculate_sum(file_path, "previous")

print(paste("Solution to part 1 is:", sum_next_values))
print(paste("Solution to part 2 is:", sum_previous_values))
